File: android/src/main/java/expo/modules/smartrefreshlayout/HeaderPropsHandler.kt
package expo.modules.smartrefreshlayout

import android.content.res.Resources
import android.view.View
import android.view.ViewGroup
import androidx.core.graphics.toColorInt
import com.scwang.smart.refresh.header.ClassicsHeader
import com.scwang.smart.refresh.layout.SmartRefreshLayout
import com.scwang.smart.refresh.layout.constant.RefreshState as NativeRefreshState

/**
 * Header 属性处理器
 * 负责处理经典样式 Header 的所有属性设置
 */
class HeaderPropsHandler(
    private val srl: SmartRefreshLayout,
    private val resources: Resources
) {
    
    /**
     * 应用 Header 属性
     */
    fun applyProps(props: Map<String, Any?>?) {
        props ?: return
        
        val header = srl.refreshHeader
        if (header !is ClassicsHeader) {
            return
        }
        
        try {
            // 设置文字内容
            setTextContent(header, props)
            
            // 强制刷新 Header 显示
            header.onStateChanged(srl, NativeRefreshState.None, srl.state)
            
            // 设置颜色
            setColors(header, props)
            
            // 设置字体大小
            setTextSize(header, props)
            
            // 设置其他属性
            setOtherProps(header, props)
            
            // 设置箭头大小
            setArrowSize(header, props)
            
            // 设置间距
            setMargin(header, props)
            
            // 监听状态变化，确保刷新图片动画运行
            setupStateListener(header, props)
        } catch (_: Throwable) {
            // 静默处理异常
        }
    }
    
    /**
     * 设置状态监听器
     * 在刷新状态时确保刷新图片动画运行
     */
    private fun setupStateListener(header: ClassicsHeader, props: Map<String, Any?>) {
        // 使用 Handler 定期检查并确保刷新图片动画运行
        val handler = android.os.Handler(android.os.Looper.getMainLooper())
        val runnable = object : Runnable {
            override fun run() {
                if (srl.state == NativeRefreshState.Refreshing) {
                    // 在刷新状态时，持续确保刷新图片动画运行
                    ensureProgressImageAnimating(header, props)
                }
                // 每 100ms 检查一次
                handler.postDelayed(this, 100)
            }
        }
        handler.post(runnable)
    }
    
    /**
     * 确保刷新图片动画运行
     * 注意：不要永久改变视图状态，让 ClassicsHeader 自己管理箭头/图片的切换
     */
    private fun ensureProgressImageAnimating(header: ClassicsHeader, props: Map<String, Any?>) {
        try {
            var searchClass: Class<*>? = ClassicsHeader::class.java
            var progressImageView: View? = null
            
            while (searchClass != null && searchClass != Any::class.java) {
                try {
                    val progressField = searchClass.getDeclaredField("mProgressView")
                    progressField.isAccessible = true
                    progressImageView = progressField.get(header) as? View
                    break
                } catch (_: NoSuchFieldException) {
                    searchClass = searchClass.superclass
                }
            }
            
            // 只确保图片动画在运行，不改变 visibility
            progressImageView?.let { imageView ->
                if (imageView is android.widget.ImageView) {
                    val drawable = imageView.drawable
                    if (drawable is android.graphics.drawable.Animatable) {
                    if (!(drawable as android.graphics.drawable.Animatable).isRunning) {
                        (drawable as android.graphics.drawable.Animatable).start()
                    }
                    }
                }
            }
        } catch (_: Exception) {
            // 静默处理异常
        }
    }
    
    /**
     * 设置文字内容
     */
    private fun setTextContent(header: ClassicsHeader, props: Map<String, Any?>) {
        val textFields = mapOf(
            "REFRESH_HEADER_PULLING" to "mTextPulling",
            "REFRESH_HEADER_REFRESHING" to "mTextRefreshing",
            "REFRESH_HEADER_LOADING" to "mTextLoading",
            "REFRESH_HEADER_RELEASE" to "mTextRelease",
            "REFRESH_HEADER_FINISH" to "mTextFinish",
            "REFRESH_HEADER_FAILED" to "mTextFailed",
            "REFRESH_HEADER_SECONDARY" to "mTextSecondary",
            "REFRESH_HEADER_UPDATE" to "mTextUpdate"
        )
        
        textFields.forEach { (propKey, fieldName) ->
            props[propKey]?.let { value ->
                if (value is String) {
                    try {
                        val field = ClassicsHeader::class.java.getDeclaredField(fieldName)
                        field.isAccessible = true
                        field.set(header, value)
                        
                        // 同时修改静态常量
                        when (propKey) {
                            "REFRESH_HEADER_PULLING" -> 
                                com.scwang.smart.refresh.header.ClassicsHeader.REFRESH_HEADER_PULLING = value
                            "REFRESH_HEADER_REFRESHING" -> 
                                com.scwang.smart.refresh.header.ClassicsHeader.REFRESH_HEADER_REFRESHING = value
                            "REFRESH_HEADER_LOADING" -> 
                                com.scwang.smart.refresh.header.ClassicsHeader.REFRESH_HEADER_LOADING = value
                            "REFRESH_HEADER_RELEASE" -> 
                                com.scwang.smart.refresh.header.ClassicsHeader.REFRESH_HEADER_RELEASE = value
                            "REFRESH_HEADER_FINISH" -> 
                                com.scwang.smart.refresh.header.ClassicsHeader.REFRESH_HEADER_FINISH = value
                            "REFRESH_HEADER_FAILED" -> 
                                com.scwang.smart.refresh.header.ClassicsHeader.REFRESH_HEADER_FAILED = value
                            "REFRESH_HEADER_SECONDARY" -> 
                                com.scwang.smart.refresh.header.ClassicsHeader.REFRESH_HEADER_SECONDARY = value
                            "REFRESH_HEADER_UPDATE" -> 
                                com.scwang.smart.refresh.header.ClassicsHeader.REFRESH_HEADER_UPDATE = value
                        }
                    } catch (_: Exception) {
                        // 静默处理异常
                    }
                }
            }
        }
    }
    
    /**
     * 设置颜色
     */
    private fun setColors(header: ClassicsHeader, props: Map<String, Any?>) {
        props["headerAccentColor"]?.let {
            if (it is String) {
                val colorStr = normalizeColorString(it)
                header.setAccentColor(colorStr.toColorInt())
            }
        }
        props["headerPrimaryColor"]?.let {
            if (it is String) {
                val colorStr = normalizeColorString(it)
                header.setPrimaryColor(colorStr.toColorInt())
            }
        }
    }
    
    /**
     * 标准化颜色字符串
     * 将 3 位十六进制颜色转换为 6 位
     * 例如：#000 -> #000000, #F00 -> #FF0000
     */
    private fun normalizeColorString(color: String): String {
        var result = color.trim()
        
        // 如果是 3 位十六进制颜色（#RGB），转换为 6 位（#RRGGBB）
        if (result.startsWith("#") && result.length == 4) {
            val r = result[1]
            val g = result[2]
            val b = result[3]
            result = "#$r$r$g$g$b$b"
        }
        
        return result
    }
    
    /**
     * 设置字体大小
     */
    private fun setTextSize(header: ClassicsHeader, props: Map<String, Any?>) {
        props["headerTitleTextSize"]?.let { 
            if (it is Number) {
                val spValue = it.toFloat()
                header.setTextSizeTitle(spValue)
            }
        }
        props["headerTimeTextSize"]?.let { 
            if (it is Number) {
                val spValue = it.toFloat()
                header.setTextSizeTime(spValue)
            }
        }
    }
    
    /**
     * 设置其他属性
     */
    private fun setOtherProps(header: ClassicsHeader, props: Map<String, Any?>) {
        props["headerShowTime"]?.let { 
            if (it is Boolean) header.setEnableLastTime(it) 
        }
        props["headerFinishDuration"]?.let { 
            if (it is Number) {
                header.setFinishDuration(it.toInt())
            }
        }
        props["headerLastUpdateText"]?.let { 
            if (it is String) header.setLastUpdateText(it) 
        }
    }
    
    /**
     * 设置箭头大小和刷新图片大小
     */
    private fun setArrowSize(header: ClassicsHeader, props: Map<String, Any?>) {
        val headerArrowSize = props["headerDrawableArrowSize"] ?: props["headerDrawableSize"]
        val progressSize = props["headerDrawableProgressSize"]
        
        // 使用 ClassicsHeader 提供的 API 设置箭头大小
        headerArrowSize?.let {
            if (it is Number) {
                val dpValue = it.toFloat()
                header.setDrawableArrowSize(dpValue)
            }
        }
        
        // 使用 ClassicsHeader 提供的 API 设置刷新图片大小
        progressSize?.let {
            if (it is Number) {
                val dpValue = it.toFloat()
                header.setDrawableProgressSize(dpValue)
            }
        }
        
        // 配置刷新图片的颜色和动画
        configureProgressImage(header, props)
    }

    
    /**
     * 配置刷新图片的颜色和动画
     * 初始化时只配置，不改变显示状态，让 ClassicsHeader 自己控制箭头/图片的切换
     */
    private fun configureProgressImage(header: ClassicsHeader, props: Map<String, Any?>) {
        try {
            var searchClass: Class<*>? = ClassicsHeader::class.java
            var progressImageView: View? = null
            
            // 查找刷新图片视图
            while (searchClass != null && searchClass != Any::class.java) {
                try {
                    val progressField = searchClass.getDeclaredField("mProgressView")
                    progressField.isAccessible = true
                    progressImageView = progressField.get(header) as? View
                    break
                } catch (_: NoSuchFieldException) {
                    searchClass = searchClass.superclass
                }
            }
            
            // 配置刷新图片（但不改变显示状态，让 ClassicsHeader 自己控制）
            progressImageView?.let { imageView ->
                // 如果是 ImageView（包括 AppCompatImageView），配置动画和颜色
                if (imageView is android.widget.ImageView) {
                    // 获取 drawable
                    val drawable = imageView.drawable
                    
                    // 启动动画
                    drawable?.let { d ->
                        if (d is android.graphics.drawable.Animatable) {
                            (d as android.graphics.drawable.Animatable).start()
                        }
                    }
                    
                    // 如果设置了强调色，应用到刷新图片
                    props["headerAccentColor"]?.let { color ->
                        if (color is String) {
                            try {
                                imageView.setColorFilter(normalizeColorString(color).toColorInt())
                            } catch (_: Exception) {
                                // 静默处理异常
                            }
                        }
                    }
                }
            }
            
        } catch (_: Exception) {
            // 静默处理异常
        }
    }
    
    /**
     * 设置间距
     */
    private fun setMargin(header: ClassicsHeader, props: Map<String, Any?>) {
        props["headerDrawableMarginRight"]?.let { 
            if (it is Number) {
                header.setDrawableMarginRight(dpToPx(it.toDouble()))
            }
        }
    }
    
    /**
     * dp 转 px
     * 
     * 将密度无关像素（dp）转换为实际像素（px）
     * 
     * @param dp 密度无关像素值
     * @return 对应的像素值
     */
    private fun dpToPx(dp: Double): Float {
        val density = resources.displayMetrics.density
        return (dp * density).toFloat()
    }
}

File: ios/HeaderPropsHandler.swift
import UIKit
import MJRefresh

/**
 * Header 属性处理器
 * 
 * 负责处理 MJRefresh Header 的所有属性设置
 * 包括：
 * - 文字内容（下拉提示、释放提示、刷新中等各种状态的文字）
 * - 颜色设置（强调色、主色调）
 * - 字体大小（标题、时间文字大小）
 * - 箭头/进度图标的显示控制
 * - 其他配置（是否显示时间、最后更新时间前缀等）
 * 
 * 注意：iOS 的 MJRefresh 对某些属性的支持有限，如箭头大小、图标间距等
 */
class HeaderPropsHandler {
    
    /**
     * 应用 Header 属性到 MJRefreshNormalHeader
     * 
     * @param header MJRefreshNormalHeader 实例
     * @param props 属性映射表
     */
    static func applyProps(header: MJRefreshNormalHeader, props: [String: Any]) {
        // 设置文字内容
        setTextContent(header: header, props: props)
        
        // 设置颜色
        setColors(header: header, props: props)
        
        // 设置字体大小
        setTextSize(header: header, props: props)
        
        // 设置其他属性
        setOtherProps(header: header, props: props)
        
        // 设置箭头显示
        setArrowVisibility(header: header, props: props)
    }
    
    /**
     * 应用 Material Header 属性
     * 
     * @param header MJRefreshMaterialHeader 实例
     * @param props 属性映射表
     */
    static func applyMaterialProps(header: MJRefreshMaterialHeader, props: [String: Any]) {
        // 设置颜色
        if let accentColor = props["headerAccentColor"] as? String,
           let color = UIColor(hexString: accentColor) {
            header.accentColor = color
        }
        
        if let primaryColor = props["headerPrimaryColor"] as? String,
           let color = UIColor(hexString: primaryColor) {
            header.headerBackgroundColor = color
        }
    }
    
    // MARK: - 私有方法
    
    /**
     * 设置文字内容
     * 
     * 设置 Header 在不同状态下显示的文字：
     * - REFRESH_HEADER_PULLING: 下拉过程中的提示文字
     * - REFRESH_HEADER_RELEASE: 释放即可刷新的提示文字
     * - REFRESH_HEADER_REFRESHING: 刷新中的提示文字
     * - REFRESH_HEADER_FINISH: 刷新完成的提示文字
     * 
     * @param header MJRefreshNormalHeader 实例
     * @param props 属性映射表
     */
    private static func setTextContent(header: MJRefreshNormalHeader, props: [String: Any]) {
        if let pullingText = props["REFRESH_HEADER_PULLING"] as? String {
            header.setTitle(pullingText, for: .idle)
            header.setTitle(pullingText, for: .pulling)
        }
        
        if let releaseText = props["REFRESH_HEADER_RELEASE"] as? String {
            header.setTitle(releaseText, for: .willRefresh)
        }
        
        if let refreshingText = props["REFRESH_HEADER_REFRESHING"] as? String {
            header.setTitle(refreshingText, for: .refreshing)
        }
        
        if let finishText = props["REFRESH_HEADER_FINISH"] as? String {
            header.setTitle(finishText, for: .noMoreData)
        }
        
        // iOS MJRefresh 不支持的状态文字（静默处理）
        _ = props["REFRESH_HEADER_LOADING"]
        _ = props["REFRESH_HEADER_FAILED"]
        _ = props["REFRESH_HEADER_SECONDARY"]
    }
    
    /**
     * 设置颜色
     * 
     * - headerAccentColor: 强调色（文字、箭头、进度条的颜色）
     * - headerPrimaryColor: 主色调（背景色，iOS MJRefresh 不支持）
     * 
     * @param header MJRefreshNormalHeader 实例
     * @param props 属性映射表
     */
    private static func setColors(header: MJRefreshNormalHeader, props: [String: Any]) {
        if let accentColor = props["headerAccentColor"] as? String,
           let color = UIColor(hexString: accentColor) {
            header.stateLabel?.textColor = color
            header.lastUpdatedTimeLabel?.textColor = color
            // iOS 的 MJRefresh 箭头颜色与文字颜色绑定
            header.arrowView?.tintColor = color
            // 设置菊花（loading indicator）的颜色
            if let loadingView = header.loadingView as? UIActivityIndicatorView {
                loadingView.color = color
            }
        }
        
        // iOS MJRefresh 不支持背景色设置
        _ = props["headerPrimaryColor"]
    }
    
    /**
     * 设置字体大小
     * 
     * - headerTitleTextSize: 标题文字大小（单位：pt）
     * - headerTimeTextSize: 时间文字大小（单位：pt）
     * 
     * @param header MJRefreshNormalHeader 实例
     * @param props 属性映射表
     */
    private static func setTextSize(header: MJRefreshNormalHeader, props: [String: Any]) {
        if let titleSize = props["headerTitleTextSize"] as? CGFloat {
            header.stateLabel?.font = UIFont.systemFont(ofSize: titleSize)
        }
        
        if let timeSize = props["headerTimeTextSize"] as? CGFloat {
            header.lastUpdatedTimeLabel?.font = UIFont.systemFont(ofSize: timeSize)
        }
    }
    
    /**
     * 设置其他属性
     * 
     * - headerShowTime: 是否显示最后更新时间
     * - headerLastUpdateText: 自定义最后更新时间的前缀文字
     * - headerFinishDuration: 刷新完成后停留时长（iOS MJRefresh 不支持）
     * 
     * @param header MJRefreshNormalHeader 实例
     * @param props 属性映射表
     */
    private static func setOtherProps(header: MJRefreshNormalHeader, props: [String: Any]) {
        // 是否显示时间
        if let showTime = props["headerShowTime"] as? Bool {
            header.lastUpdatedTimeLabel?.isHidden = !showTime
        }
        
        // iOS MJRefresh 不支持的属性（静默处理）
        _ = props["headerLastUpdateText"]
        _ = props["headerFinishDuration"]
    }
    
    /**
     * 设置箭头显示
     * 
     * 支持两个属性（优先使用第一个）：
     * - headerDrawableArrowSize: 专门设置箭头大小
     * - headerDrawableSize: 通用图标大小设置
     * 
     * 特殊值：
     * - 0: 隐藏箭头
     * - 其他值: iOS MJRefresh 不支持动态调整箭头大小，仅记录日志
     * 
     * @param header MJRefreshNormalHeader 实例
     * @param props 属性映射表
     */
    private static func setArrowVisibility(header: MJRefreshNormalHeader, props: [String: Any]) {
        // 箭头大小设置（优先使用 headerDrawableArrowSize，回退到 headerDrawableSize）
        let arrowSize = props["headerDrawableArrowSize"] as? CGFloat ?? props["headerDrawableSize"] as? CGFloat
        
        if let size = arrowSize {
            // 如果设置为 0，隐藏箭头
            if size == 0 {
                header.arrowView?.isHidden = true
            } else {
                header.arrowView?.isHidden = false
                // iOS 的 MJRefresh 箭头大小是固定的，无法直接修改
            }
        }
        
        // iOS MJRefresh 不支持的属性（静默处理）
        _ = props["headerDrawableProgressSize"]
        _ = props["headerDrawableMarginRight"]
    }
}
